crypto/fipsmodule/sha/internal.h

```diff
@@ -406,39 +409,91 @@ OPENSSL_EXPORT uint8_t *SHAKE128(const uint8_t *data, const size_t in_len,
+// SHAKE_Init initialises |ctx| fields through |FIPS202_Init| and
+// returns 1 on success and 0 on failure. SHAKE_Init fails if
+// |ctx| is nullptr or |block_size| is inproper for SHAKE128 or
+// SHAKE256. Otherwise, it propagates the |FIPS202_Init| return value.
 int SHAKE_Init(KECCAK1600_CTX *ctx, size_t block_size);
 
-// SHAKE_Absorb checks |ctx| pointer and |len| values. It updates and absorbs 
-// input blocks via |FIPS202_Update|.
+// SHAKE_Absorb checks |ctx| pointer and |len| values. It updates and absorbs
+// input blocks via |FIPS202_Update|. SHAKE_Absorb fails if |ctx| is nullptr
+// or |data| is nullptr and |len| is different from 0. Otherwise, it
+// propagates the |FIPS202_Update| return value.
 int SHAKE_Absorb(KECCAK1600_CTX *ctx, const void *data,
                                size_t len);
 
-// SHAKE_Squeeze pads the last data block and absorbs it through 
-// |FIPS202_Finalize| on first call. It writes |len| bytes of incremental 
-// XOF output to |md| and returns 1 on success and 0 on failure. It can be 
-// called multiple times.
+// SHAKE_Squeeze pads the last data block and absorbs it through
+// |FIPS202_Finalize| on first call. It writes |len| bytes of incremental
+// XOF output to |md| and returns 1 on success and 0 on failure. It can be
+// called multiple times. SHAKE_Squeeze fails if |ctx| or |md| are nullptr,
+// when called from inappropriate |ctx->state|, or on |FIPS202_Finalize| fail.
+// Otherwise, it returns 1.
 int SHAKE_Squeeze(uint8_t *md, KECCAK1600_CTX *ctx, size_t len);
 
 // SHAKE_Final writes |len| bytes of finalized extendible output to |md|, returns 1 on
 // success and 0 on failure. It should be called once to finalize absorb and
 // squeeze phases. Incremental XOF output should be generated via |SHAKE_Squeeze|.
+// SHAKE_Final fails if |ctx| or |md| are nullptr or on |FIPS202_Finalize| fail.
+// Otherwise, it returns 1.
 int SHAKE_Final(uint8_t *md, KECCAK1600_CTX *ctx, size_t len);
 
+// SHAKE128_Init_x4 is a batched API that operates on four independent
+// Keccak bitstates. It initialises all four |ctx| fields through four
+// consecutive calls to |SHAKE_Init| and returns 1 on success and 0 on failure.
+// SHAKE128_Init_x4 succeeds when all four |SHAKE_Init| functions succeed.
+// It fails on the first |SHAKE_Init| function fail, however, if called with
+// valid inputs, this function never fails.
+OPENSSL_EXPORT int SHAKE128_Init_x4(KECCAK1600_CTX_x4 *ctx);
+
+// SHAKE128_Absorb_once_x4 is a batched API that operates on four independent
+// Keccak bitstates. It absorbs all four inputs |data0|, |data1|, |data2|, |data3|
+// of equal length of |len| bytes through four consecutive calls to |SHAKE_Absorb|
+// and returns 1 on success and 0 on failure. SHAKE128_Absorb_once_x4 succeeds when
+// all four |SHAKE_Absorb| functions succeed. It fails on the first |SHAKE_Absorb|
+// function fail, however, if called with valid inputs, this function never fails.
```

Comment:
It is unclear what 'valid' means here. You want the input to come fresh out of `SHAKE128_Init_x4`, or?

Reply:
'Valid input' -> 'valid input pointer'. 
However, I updated this following the suggested documentation format. Thank you.